File: src/api/new-discussion-recipients.ts
import { RequestFN } from "~/core/request-function";
import { decodeNewDiscussionRecipient } from "~/decoders/new-discussion-recipient";
import { type SessionHandle, TabLocation, type EntityKind, type NewDiscussionRecipient } from "~/models";

/**
 * Returns a list of possible recipients when creating a discussion.
 *
 * This step is required before creating a discussion.
 * It allows to know who can be the recipient of the discussion.
 *
 * @param kind The kind of entity to create a discussion with. Only `Teacher`, `Student` and `Personal` are allowed.
 */
export const newDiscussionRecipients = async (session: SessionHandle, kind: EntityKind): Promise<Array<NewDiscussionRecipient>> => {
  const user = session.userResource;

  // TODO: use `ListePublics` for teachers.
  const request = new RequestFN(session, "ListeRessourcesPourCommunication", {
    _Signature_: { onglet: TabLocation.Discussions },

    donnees: {
      filtreElement: {
        G: user.kind,
        L: user.name,
        N: user.id
      },

      onglet: {
        N: 0,
        G: kind
      }
    }
  });

  const response = await request.send();
  return response.data.donnees.listeRessourcesPourCommunication.V
    .filter((recipient: any) => recipient.avecDiscussion)
    .map(decodeNewDiscussionRecipient);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/new-discussion-recipients.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeNewDiscussionRecipient
import ink.literate.pawnote.models.EntityKind
import ink.literate.pawnote.models.NewDiscussionRecipient
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

/**
 * Returns a list of possible recipients when creating a discussion.
 *
 * This step is required before creating a discussion. It allows to know who can be the recipient of
 * the discussion.
 *
 * @param kind The kind of entity to create a discussion with. Only `Teacher`, `Student` and
 *   `Personal` are allowed.
 */
suspend fun newDiscussionRecipients(
    session: SessionHandle,
    kind: EntityKind
): List<NewDiscussionRecipient> {
  val user = session.userResource

  // TODO: use `ListePublics` for teachers.
  val request =
      RequestFN(
          session.information,
          "ListeRessourcesPourCommunication",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Discussions.code) }

                putJsonObject("donnees") {
                  putJsonObject("filtreElement") {
                    put("G", user.kind)
                    put("L", user.name)
                    put("N", user.id)
                  }

                  putJsonObject("onglet") {
                    put("N", 0)
                    put("G", kind.code)
                  }
                }
              }))

  val response = request.send()
  return Json.parseToJsonElement(response.data)
      .jsonObject["donnees"]!!
      .jsonObject["listeRessourcesPourCommunication"]!!
      .jsonObject["V"]!!
      .jsonArray
      .filter { it.jsonObject["avecDiscussion"]?.jsonPrimitive?.boolean ?: false }
      .map { decodeNewDiscussionRecipient(it.jsonObject) }
}
